packages/core/src/tools/templating/utils.ts

```diff
@@ -100,3 +101,107 @@ export function getItems<Item>(settings: GetItemsSettings<Item>): Item[] {
 
   return items;
 }
+
+// ---- Low-level grouping (returns raw groups, NOT entries)
+
+/**
+ * Converts a given path to POSIX format by replacing backslashes with forward slashes.
+ * @param p
+ */
+function toPosix(p: string) {
+  return p.replace(/\\/g, '/');
+}
+
+/**
+ * Groups components by directory and base component name (prefix).
+ * For example, "Teaser" and "Teaser.Variant1" in the same folder are grouped together.
+ * @param {ComponentSource[]} files list of components
+ * @returns {ComponentGroup[]} list of groups
+ */
+export function groupComponentsByDirAndPrefix<T extends ComponentSource>(
+  files: T[]
+): ComponentGroup<T>[] {
+  const groupsByKey = new Map<string, ComponentGroup<T>>();
+
+  const getBaseComponentName = (componentName: string) => {
+    const dotIndex = componentName.indexOf('.');
+    return dotIndex === -1 ? componentName : componentName.slice(0, dotIndex);
+  };
+
+  for (const file of files) {
+    if (!file || !file.filePath || !file.componentName) continue;
+
+    const posixPath = toPosix(file.filePath);
+    if (!posixPath) continue;
+    const directory = path.posix.dirname(posixPath);
+    const prefix = getBaseComponentName(file.componentName);
+    const compositeKey = `${directory}::${prefix}`;
+
+    let group = groupsByKey.get(compositeKey);
+    if (!group) {
+      group = { dir: directory, prefix, neighbors: [] };
+      groupsByKey.set(compositeKey, group);
+    }
+
+    if (file.componentName === prefix) {
+      group.base = file;
+    } else {
+      group.neighbors.push(file);
+    }
+  }
+
+  const componentGroups = Array.from(groupsByKey.values());
+
+  return componentGroups;
+}
+
+/**
+ * Produces entries where each base component key aggregates its variants via object spread.
+ * @param {ComponentSource[]} components list of components
+ * @returns {ComponentMapEntry[]} list of map entries
+ */
+export function groupComponentsWithVariants(components: ComponentSource[]): ComponentMapEntry[] {
```

Comment:
maybe this can be simplified and this function can be cut too.
i.e. in generateMap we can always look for variants, group regular+variant component files together and then we'll have a list of components that are either:
- grouped with a variant
- just present on their own
groups will get a spread treatment, and regular components will just be as they are.
You already do a similar logic in buildNextMapContent - and I will outline my prosposal more in my next comment for that function